### ISSUE TITLE
terminate called without an active exception terminate called recursively terminate called recursively

### ISSUE THREAD

**Issue by eliassama** [qdrant/qdrant#4195]
Our service has been running fine, however, it has recently started to experience the following problem:

![image](https://github.com/qdrant/qdrant/assets/79587688/35e21de8-6cc3-4842-846b-38ec0bbd35f3)

It will appear every time we restart

![image](https://github.com/qdrant/qdrant/assets/79587688/ee066192-9cf1-45c0-8d64-f55293fe26c2)

From docker, he is running fine, but cannot be accessed through any port

![image](https://github.com/qdrant/qdrant/assets/79587688/a927a6b3-5075-4024-b137-2a8c03653978)

**Comment by eliassama** [comment]
By the way, we only have 33 collections。This one looks similar to [#3564](https://github.com/qdrant/qdrant/issues/3564), but not the same?

**Comment by eliassama** [comment]
My memory usage is high, is this related to it?

**Comment by eliassama** [comment]
I doubled the memory, but it still doesn't work. Let me summarize:

Problem performance:
1. Log prompt: terminate called without an active exception
terminate called recursively
2. Reading and writing operations cannot be performed normally.
3. Port access 6333 6334 6335 prompts: Connection refused
4. The dashboard cannot be accessed normally. This should be because when accessing port 6333, a Connection refused message is displayed.

I did the following:
1. Restart the service.
2. Double the memory upgrade.

I now need to understand the cause of this problem and how to get my service back to normal.

**Comment by eliassama** [comment]
@timvisee Hello, can you help me solve this problem?

**Comment by agourlay** [comment]
Have you checked if you are not running of memory while starting up?

**Comment by timvisee** [comment]
@eliassama I am quite puzzled by this and am not able to reproduce this yet.

Are you able to simplify your test case that reproduces it, so I can try to match that?

**Comment by eliassama** [comment]
> @eliassama I am quite puzzled by this and am not able to reproduce this yet.
> 
> Are you able to simplify your test case that reproduces it, so I can try to match that?

There is no way for me to reproduce it on the smallest new cluster. I can provide my data volume, version and configuration file.

data amount: 429G 

## compose
```yaml
version: '3.5'

services:
  qdrant:
    container_name: qdrant
    image: qdrant/qdrant:v1.6.0
    environment:
      - QDRANT__CLUSTER__ENABLED=true
    tty: true
    ports:
      - "6333:6333"
      - "6334:6334"
      - "6335:6335"
    volumes:
      - /disk/ssd1/qdrant-node/storage:/disk/storage
      - /disk/ssd1/qdrant-node/snapshots:/disk/snapshots
      - /disk/ssd1/qdrant-node/config/production.yaml:/qdrant/config/production.yaml
    command: /qdrant/qdrant --uri http://qdrantHost:6335
    networks:
     - qrant

networks:
  qrant:
    driver: bridge
    ipam:
      driver: default
      config:
      - subnet: 21.88.12.0/24
        gateway: 21.88.12.1
```

## config

```yaml
log_level: INFO

storage:
  # Where to store all the data
  storage_path: /disk/storage

  # Where to store snapshots
  snapshots_path: /disk/snapshots

  # Where to store temporary files
  # If null, temporary snapshot are stored in: storage/snapshots_temp/
  temp_path: /disk/snapshots_temp

  # If true - point's payload will not be stored in memory.
  # It will be read from the disk every time it is requested.
  # This setting saves RAM by (slightly) increasing the response time.
  # Note: those payload values that are involved in filtering and are indexed - remain in RAM.
  on_disk_payload: true

  # Maximum number of concurrent updates to shard replicas
  # If `null` - maximum concurrency is used.
  update_concurrency: null

  # Write-ahead-log related configuration
  wal:
    # Size of a single WAL segment
    wal_capacity_mb: 32

    # Number of WAL segments to create ahead of actual data requirement
    wal_segments_ahead: 0

  # Normal node - receives all updates and answers all queries
  node_type: "Normal"

  # Listener node - receives all updates, but does not answer search/read queries
  # Useful for setting up a dedicated backup node
  # node_type: "Listener"

  performance:
    # Number of parallel threads used for search operations. If 0 - auto selection.
    max_search_threads: 0
    # Max total number of threads, which can be used for running optimization processes across all collections.
    # Note: Each optimization thread will also use `max_indexing_threads` for index building.
    # So total number of threads used for optimization will be `max_optimization_threads * max_indexing_threads`
    max_optimization_threads: 1

  optimizers:
    # The minimal fraction of deleted vectors in a segment, required to perform segment optimization
    deleted_threshold: 0.2

    # The minimal number of vectors in a segment, required to perform segment optimization
    vacuum_min_vector_number: 1000

    # Target amount of segments optimizer will try to keep.
    # Real amount of segments may vary depending on multiple parameters:
    #  - Amount of stored points
    #  - Current write RPS
    #
    # It is recommended to select default number of segments as a factor of the number of search threads,
    # so that each segment would be handled evenly by one of the threads.
    # If `default_segment_number = 0`, will be automatically selected by the number of available CPUs
    default_segment_number: 0

    # Do not create segments larger this size (in KiloBytes).
    # Large segments might require disproportionately long indexation times,
    # therefore it makes sense to limit the size of segments.
    #
    # If indexation speed have more priority for your - make this parameter lower.
    # If search speed is more important - make this parameter higher.
    # Note: 1Kb = 1 vector of size 256
    # If not set, will be automatically selected considering the number of available CPUs.
    max_segment_size_kb: null

    # Maximum size (in KiloBytes) of vectors to store in-memory per segment.
    # Segments larger than this threshold will be stored as read-only memmaped file.
    # To enable memmap storage, lower the threshold
    # Note: 1Kb = 1 vector of size 256
    # To explicitly disable mmap optimization, set to `0`.
    # If not set, will be disabled by default.
    memmap_threshold_kb: 10000

    # Maximum size (in KiloBytes) of vectors allowed for plain index.
    # Default value based on https://github.com/google-research/google-research/blob/master/scann/docs/algorithms.md
    # Note: 1Kb = 1 vector of size 256
    # To explicitly disable vector indexing, set to `0`.
    # If not set, the default value will be used.
    indexing_threshold_kb: 20000

    # Interval between forced flushes.
    flush_interval_sec: 5

    # Max number of threads, which can be used for optimization per collection.
    # Note: Each optimization thread will also use `max_indexing_threads` for index building.
    # So total number of threads used for optimization will be `max_optimization_threads * max_indexing_threads`
    # If `max_optimization_threads = 0`, optimization will be disabled.
    max_optimization_threads: 1

  # Default parameters of HNSW Index. Could be overridden for each collection or named vector individually
  hnsw_index:
    # Number of edges per node in the index graph. Larger the value - more accurate the search, more space required.
    m: 32
    # Number of neighbours to consider during the index building. Larger the value - more accurate the search, more time required to build index.
    ef_construct: 100
    # Minimal size (in KiloBytes) of vectors for additional payload-based indexing.
    # If payload chunk is smaller than `full_scan_threshold_kb` additional indexing won't be used -
    # in this case full-scan search should be preferred by query planner and additional indexing is not required.
    # Note: 1Kb = 1 vector of size 256
    full_scan_threshold_kb: 10000
    # Number of parallel threads used for background index building. If 0 - auto selection.
    max_indexing_threads: 0
    # Store HNSW index on disk. If set to false, index will be stored in RAM. Default: false
    on_disk: true
    # Custom M param for hnsw graph built for payload index. If not set, default M will be used.
    payload_m: 32


service:

  # Maximum size of POST data in a single request in megabytes
  max_request_size_mb: 32

  # Number of parallel workers used for serving the api. If 0 - equal to the number of available cores.
  # If missing - Same as storage.max_search_threads
  max_workers: 0

  # Host to bind the service on
  host: 0.0.0.0

  # HTTP(S) port to bind the service on
  http_port: 6333

  # gRPC port to bind the service on.
  # If `null` - gRPC is disabled. Default: null
  # Comment to disable gRPC:
  grpc_port: 6334

  # Enable CORS headers in REST API.
  # If enabled, browsers would be allowed to query REST endpoints regardless of query origin.
  # More info: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
  # Default: true
  enable_cors: true

  # Enable HTTPS for the REST and gRPC API
  enable_tls: false

  # Check user HTTPS client certificate against CA file specified in tls config
  verify_https_client_certificate: false

  # Set an api-key.
  # If set, all requests must include a header with the api-key.
  # example header: `api-key: <API-KEY>`
  #
  # If you enable this you should also enable TLS.
  # (Either above or via an external service like nginx.)
  # Sending an api-key over an unencrypted channel is insecure.
  #
  # Uncomment to enable.
  # api_key: your_secret_api_key_here

cluster:
  # Use `enabled: true` to run Qdrant in distributed deployment mode
  enabled: true

  # Configuration of the inter-cluster communication
  p2p:
    # Port for internal communication between peers
    port: 6335

    # Use TLS for communication between peers
    enable_tls: false

  # Configuration related to distributed consensus algorithm
  consensus:
    # How frequently peers should ping each other.
    # Setting this parameter to lower value will allow consensus
    # to detect disconnected nodes earlier, but too frequent
    # tick period may create significant network and CPU overhead.
    # We encourage you NOT to change this parameter unless you know what you are doing.
    tick_period_ms: 100


# Set to true to prevent service from sending usage statistics to the developers.
# Read more: https://qdrant.tech/documentation/guides/telemetry
telemetry_disabled: false


# TLS configuration.
# Required if either service.enable_tls or cluster.p2p.enable_tls is true.
tls:
  # Server certificate chain file
  cert: ./tls/cert.pem

  # Server private key file
  key: ./tls/key.pem

  # Certificate authority certificate file.
  # This certificate will be used to validate the certificates
  # presented by other nodes during inter-cluster communication.
  #
  # If verify_https_client_certificate is true, it will verify
  # HTTPS client certificate
  #
  # Required if cluster.p2p.enable_tls is true.
  ca_cert: ./tls/cacert.pem

  # TTL in seconds to reload certificate from disk, useful for certificate rotations.
  # Only works for HTTPS endpoints. Does not support gRPC (and intra-cluster communication).
  # If `null` - TTL is disabled.
  cert_ttl: 3600
```

**Comment by eliassama** [comment]
@timvisee I see that the latest version is 1.9.1. Is it possible to solve this problem if we directly upgrade to the latest version?

**Comment by timvisee** [comment]
You can upgrade, but I don't expect it to solve the situation.

By the way, I see that you're creating a lot of collections. We highly recommend not to do this and to use a different approach instead. That's described here: https://qdrant.tech/documentation/guides/multiple-partitions/

Though Qdrant should not crash when doing this, I do assume that changing like describe above fixes the problem.

**Comment by eliassama** [comment]
> You can upgrade, but I don't expect it to solve the situation.
> 
> By the way, I see that you're creating a lot of collections. We highly recommend not to do this and to use a different approach instead. That's described here: https://qdrant.tech/documentation/guides/multiple-partitions/
> 
> Though Qdrant should not crash when doing this, I do assume that changing like describe above fixes the problem.

How many sets does it mean? We only have 33 sets in total, and the number of sets is fixed.

In other words, can a collection carry 400+G of data on a single machine?

**Comment by eliassama** [comment]
> You can upgrade, but I don't expect it to solve the situation.
> 
> By the way, I see that you're creating a lot of collections. We highly recommend not to do this and to use a different approach instead. That's described here: https://qdrant.tech/documentation/guides/multiple-partitions/
> 
> Though Qdrant should not crash when doing this, I do assume that changing like describe above fixes the problem.

Regarding multi-tenancy, we have made relevant configurations.

We have about 3 million users and tens of thousands of user queries in parallel, so we divide the 3 million users into 33 collections.

**Comment by eliassama** [comment]
@timvisee 

Hello, I'm here again. Now I'm trying to use only one collection to store and search data through multi-tenancy.

But when I wrote to 2/3, I found that there was a problem with the sharding status of this collection, but when I checked the cluster status and restored the cluster, it was normal.

The version I am using now is 1.9.1

<img width="1408" alt="image" src="https://github.com/qdrant/qdrant/assets/79587688/654831e2-6777-4e18-a8c6-e60d25bf5145">

<img width="995" alt="image" src="https://github.com/qdrant/qdrant/assets/79587688/a12383a4-556b-4f0c-b249-20d73a2129d9">

<img width="1154" alt="image" src="https://github.com/qdrant/qdrant/assets/79587688/33319313-4239-4425-834e-c9dd97cd314b">

**Comment by eliassama** [comment]
The error reported in the log is as follows

<img width="1495" alt="image" src="https://github.com/qdrant/qdrant/assets/79587688/4ad2b2e2-54ba-4125-9216-78fe5da1048b">

**Comment by eliassama** [comment]
How should I restore these shards

**Comment by eliassama** [comment]
@timvisee 

I think I can reproduce the problem, which refers to the terminate called recursively error.

I rebuilt a three-machine sharded cluster with version 1.9.1.

I re-imported the previous data into the three-slice cluster and I used the following method.

1. Write to three machines on average.
2. Each batch of 1,000 pieces of data is written, and approximately 5,000 - 10,000 pieces of data are written every second, that is, 5 - 10 batches of data are written in parallel.

Then after a night like this, shard dead and shard partial appeared.

I think the reason for this problem is that the processing speed of qdrant cannot keep up with the writing speed, that is, the writing traffic is too large, causing the internal sharding to collapse.

This has nothing to do with whether there are multiple collections, it is purely due to the excessive amount of writing.

Is there any good fix for qdrant in this regard?

<img width="717" alt="image" src="https://github.com/qdrant/qdrant/assets/79587688/c4d855d8-961c-4a65-90e8-ff95a5f19cee">

**Comment by eliassama** [comment]
I think an excellent vector database can support high concurrent writes.

If the amount of writing is too large, you can directly reject the latest written data and return an exception error. In this case, the problem will be easier to troubleshoot.

And we may only need to add a retry mechanism, or use a message queue to cut peaks and fill valleys.

**Comment by generall** [comment]
i don't think this is related to concurrent writes. We do have a mechanism to limit the writes and the stack overflow happens on restart

**Comment by eliassama** [comment]
> i don't think this is related to concurrent writes. We do have a mechanism to limit the writes and the stack overflow happens on restart

At present, the shard dead and shard partial issues occur, if not restored for a long time, restart will cause terminate called recursively

The whole cluster was rebuilt using the new version 1.9.1, and after the cluster was rebuilt, the only thing I did was batch write.

So, I think that's a reasonable explanation.

As for why the restriction did not work and the recovery did not succeed, perhaps it was only in certain scenarios? For example, using my configuration and collection parameters to create it might appear?

**Comment by eliassama** [comment]
This is my collection parameter, and my configuration file is posted above. Does the combination of the two, in certain circumstances, lead to an anomaly that cannot be recovered?

```json
PUT collections/photos
{
  "vectors": {
    "size": 512,
    "distance": "Cosine",
    "hnsw_config": {
      "m": 32,
      "ef_construct": 100,
      "full_scan_threshold": 10000,
      "max_indexing_threads": 0,
      "on_disk": true,
      "payload_m": 32
    },
    "quantization_config": {
      "scalar": {
        "type": "int8",
        "quantile": 1,
        "always_ram": false
      }
    },
    "on_disk": true
  },
  "shard_number": 3,
  "replication_factor": 3,
  "write_consistency_factor": 2,
  "on_disk_payload": true,
  "hnsw_config": {
    "m": 32,
    "ef_construct": 100,
    "full_scan_threshold": 10000,
    "max_indexing_threads": 0,
    "on_disk": true,
    "payload_m": 32
  },
  "wal_config": {
    "wal_capacity_mb": 21,
    "wal_segments_ahead": 0
  },
  "optimizers_config": {
    "deleted_threshold": 0.2,
    "vacuum_min_vector_number": 1000,
    "default_segment_number": 0,
    "max_segment_size": 0,
    "memmap_threshold": 10000,
    "indexing_threshold": 20000,
    "flush_interval_sec": 5,
    "max_optimization_threads": 1
  },
  "init_from": null,
  "quantization_config": {
      "scalar": {
        "type": "int8",
        "quantile": 1,
        "always_ram": false
      }
    }
}

```

**Comment by eliassama** [comment]
By the way, our data volume is relatively large, for example, when I imported from the backup database this time, I wrote about 20 million vector data an hour.

It took us more than ten hours of writing to recover less than half the data.

In addition, each vector data is 512 dimensions, with three payload parameters.

So, I suspect it's a problem with large volumes of writing.

In addition, our disk is using ssd, we have tried to use mechanical hard disk before, and found that it was completely unable to meet expectations.

**Comment by eliassama** [comment]
@timvisee @generall 

In addition, I would like to ask whether qdrant supports master-slave clustering.

Just like mysql or redis, there are master and slave nodes.

If the primary cluster fails, you can switch to the secondary cluster.

**Comment by generall** [comment]
in qdrant there is no difference between replicas

**Comment by eliassama** [comment]
Can replicas be run as separate read-only nodes?

**Comment by timvisee** [comment]
> Can replicas be run as separate read-only nodes?
> 

Not right now, but that is in the works. You can limit writes to it yourself externally.

I see you closed the issue. Did you manage to solve this issue?

**Comment by eliassama** [comment]
> > Can replicas be run as separate read-only nodes?
> 
> Not right now, but that is in the works. You can limit writes to it yourself externally.
> 
> I see you closed the issue. Did you manage to solve this issue?

Yes, I did the following:
1. Limit write rate
2. Increase memory
3. Upgrade to the new version 1.9.1 (but this step may be useless)

So far everything looks fine.

**Comment by eliassama** [comment]
> > > Can replicas be run as separate read-only nodes?
> > 
> > 
> > Not right now, but that is in the works. You can limit writes to it yourself externally.
> > I see you closed the issue. Did you manage to solve this issue?
> 
> Yes, I did the following:
> 
> 1. Limit write rate
> 2. Increase memory
> 3. Upgrade to the new version 1.9.1 (but this step may be useless)
> 
> So far everything looks fine.

In addition, 33 collections are merged into 1

**Comment by Zeeshan0123** [comment]
@eliassama please provide the code which you use to reduce the write rate. Thanks
